### DISCUSSION TITLE
Difference between `*` and `<` in mfshdd.cfg?

### DISCUSSION THREAD

**Discussion by NodeGuy** [moosefs/moosefs#430] [Q&A]
[The documentation](https://github.com/moosefs/moosefs/blob/master/mfsdata/mfshdd.cfg) reads:

>  - '*' means that this hard drive is 'marked for removal' and all data will be replicated to other hard drives (usually on other chunkservers)
>  - '<' means that all data from this hard drive should be moved to other local hard drives

These options seem similar to me.  What's the difference between the two?

**Comment by Blackpaw** [comment]
"*" Means the chunks will be distributed across *all* chunkservers

"<" means the chunks will be distributed amongst the remain drives on the *same* chunkserver.

**Reply by chogata** [reply]
You are right, however, it's important to remember another difference in behaviour of those two:
- with "*" chunks will always be copied to another CS and will NOT be deleted from the marked drive
- with "<" chunks will be copied to another local disk (if ">" is present too, then in first place those disks will be used) and the source copy will be deleted.

"*" is mostly designed for disks that are about to fail and you want to make copies of all chunks that are on it and you don't even care to use them as source that much (if another copy is available somewhere else), you only want your MooseFS to have one extra copy of those chunks, so that when you finally pull this drive, you will have no undergoals

"<" and ">" are designed for replacing one disk with another one, possible because the first one is failing, but not so much so that read operations are a problem (yet) or because you want to put in a faster or bigger one; you don't want replications to bother the network, you want to copy everything locally

**Reply by NodeGuy** [reply]
Thank you.  It still doesn't quite make sense to me.

I understand that "*" doesn't delete chunks and "<" does.

However, in both cases of a disk about to fail or replacing a disk with another one, what I care about is that when I pull the disk I won't have any undergoals.  Why do I care where the chunks go, i.e., why doesn't "\*" send copies of the chunks to another local disk instead of bothering the network?  I'm having trouble undertanding why I would ever use "*" instead of "<".

**Reply by Blackpaw** [reply]
In the case where there is insufficient space on the local chunkserver to store the data or where you need to decommission the entire chunkserver.

**Reply by chogata** [reply]
As Blackpaw said, if there is no space on the local chunkserver to copy all the chunks, you need to use "*" anyway. But you may need to use it even if you do have the space.

When you use "<", copies will be read from the marked disk and then deleted from the marked disk (it will receive i/o for each copy, both read only i/o and modifying i/o, because deleting a chunk means a write operation, even, if only a small one).

When you use "*", copies will be read from the whole instance and actually copies from other disks than the marked one will be preferred as source for replication. Copies will not be deleted just because they've been copied (they will be deleted if all copies need to be deleted, i.e. because someone deleted a file from MooseFS). So on average this disk will receive much less i/o than a disk with "<", which may be important for a disk that shows serious signs of failure.

One may ask, if you have a disk with serious signs of failure, why not just pull it, have your instance with undergoals for a while and let it replicate what it needs? Well, that is a solution too. It all depends on how your system is configured. If you have at least 3 copies of each chunk, then probably pulling a seriously failing disk is a better solution. But if you only have 2, pulling a disk means a portion of your chunks will be endangered and any other failure in the system may render some files unreadable. Then it's better to use the "*" and make sure the chunks are copied first.

**Reply by NodeGuy** [reply]
I see, thank you for the explanation.

So would it be accurate to say that when I want to pull a disk from a chunkserver, I should ask myself "Is the disk failing?"  

If the answer is "yes" then I should chose "*" to minimize I/O on that disk.
If the answer is "no" then I should choose "<" to minimize I/O on the network.

**Reply by chogata** [reply]
Yes, that's one of the deciding factors for sure. As I said before, sometimes you just pick "*", because you don't have the space on the local chunkserver anyway or you want to retire the whole machine. But in terms of replacing one disk, this is the question you should ask yourself to help you choose the method.

**Comment by rusticek** [comment]
e.g i have 4 chunkservers, goal 2
it is possible to mark "*" one hdd from each chunkserver without making some chunks unavailable?

I tried this and some chunks were showing as missing at look to _Regular chunks state matrix_, but in _All chunks state matrix_ everything was looking good.

**Reply by chogata** [reply]
Yes, it's possible to mark more than one hdd in a system with "*". You just need to remember there needs to be enough free space in the system as a whole to copy all the chunks from the marked hdds to other drives.

It does not make any chunks unavailable. If a chunk has a copy on disks other than those marked with "*", then this copy will be selected for read and replicate (as source) operations. If all copies of a chunk happen to sit on disks marked with "*", then one copy at random will be selected for read and replicate (as source) operations. If there is a write (modification) operation for such a chunk, all copies will get updated, no matter if they are on regular disks or disks marked with "*". Once there are enough copies of all chunks from a disk marked with "*" on other disks, this disk will be marked as ready to be removed.

The CGI tabs are there to help you gauge "what would happen if I pulled all disks I marked with * right now". If you did that, what you see on the _Regular chunks state matrix_ would be your actual situation in the system. But it's not, it's what you see in _All chunks state matrix_ that is your reality. But when the _Regular chunks state matrix_ no longer shows missing/endangered/undergoal chunks, then you know you are ready to pull you marked drives. Some users also decide to pull them when there are no missings and no endangereds and the number of undergoals is low. It depends on the circumstances (aka why you marked your disks with *, were they failing or did you just want to empty them etc.).

So to sum it up: marking disks with "*" never makes any chunks unavailable, it just changes the system's behaviour: what to replicate and which copy to use for read operations. Also prepares the system in such a way, that the disks marked with "*" can be pulled without endangering your data.
